### PR TITLE
hydra/mpiexec: refactor pg and proxy_list

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Use array to manage `HYD_pg` and `HYD_proxy`, thus `(pgid, proxy_id)` consistently identifies proxy between server and proxy processes.

Misc cleanup and preparation for reuse a node proxy for spawning processes. Ref. https://github.com/pmodels/mpich/pull/5943

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/pmi
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/pmi
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/pmi
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/pmi
test:mpich/ch3/most
test:mpich/ch4/most

File: src/pm/hydra/mpiexec/proxy.c [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "hydra_server.h"
+
+static HYD_status alloc_proxy(struct HYD_proxy **proxy, int pgid, struct HYD_node *node);
+static HYD_status add_exec_to_proxy(struct HYD_exec *exec, struct HYD_proxy *proxy, int num_procs);
+
+HYD_status HYDU_create_proxy_list(int count, struct HYD_exec *exec_list,
+                                  struct HYD_node *node_list, int pgid, bool is_singleton,
```
Comment (raffenet) [inline src/pm/hydra/mpiexec/proxy.c]: Did this file get added by mistake?

File: src/pm/hydra/mpiexec/proxy.c [hunk 2]
```diff
@@ -1,256 +0,0 @@
...
- */
-
-#include "hydra_server.h"
-
-static HYD_status alloc_proxy(struct HYD_proxy **proxy, int pgid, struct HYD_node *node);
-static HYD_status add_exec_to_proxy(struct HYD_exec *exec, struct HYD_proxy *proxy, int num_procs);
-
-HYD_status HYDU_create_proxy_list(int count, struct HYD_exec *exec_list,
-                                  struct HYD_node *node_list, int pgid, bool is_singleton,
-                                  int *proxy_count_p, struct HYD_proxy **proxy_list_p)
```
Comment (raffenet) [inline src/pm/hydra/mpiexec/proxy.c]: Yes, does look like it was added by mistake then removed 😄.

**Review by raffenet — APPROVED** [review]
Looks good. There's a stray file committed, then removed, but otherwise I don't see any showstoppers.

File: src/pm/hydra/mpiexec/proxy.c (hunk 2 above)
Comment (hzhou) [reply src/pm/hydra/mpiexec/proxy.c]: Good catch! I cleaned up before merge.

**Pull request merged**